homeassistant/components/sensor/__init__.py

```diff
@@ -267,15 +268,36 @@ def unit_of_measurement(self) -> str | None:
 
     @final
     @property
-    def state(self) -> Any:
+    def state(self) -> StateType:
         """Return the state of the sensor and perform unit conversions, if needed."""
         unit_of_measurement = self.native_unit_of_measurement
         value = self.native_value
+        device_class = self.device_class
+
+        if (
+            value is not None
+            and device_class in (DEVICE_CLASS_DATE, DEVICE_CLASS_TIMESTAMP)
+            and not isinstance(value, (datetime, date))
+        ):
+            try:
+                ciso8601.parse_datetime(str(value))
+            except (ValueError, IndexError) as error:
+                raise ValueError(
+                    f"Invalid date/datetime: {self.entity_id} provide state '{value}', "
+                    f"while it has device class '{device_class}'"
+                ) from error
+
+        if value is not None and isinstance(value, (datetime, date)):
```

Comment:
I think we can avoid the isinstance check and check the device class first, then try to call `isoformat` and raise if that fails. The isinstance check is a bit expensive as I understand it.

It will let other cases through but that should be ok, since we haven't defined types for sensors that don't have a device class.